### PR TITLE
refactor(storage): resumable uploads to use Rest library

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9274)

### CODE REVIEW

File: google/cloud/storage/internal/rest_client.cc [hunk 1]
```diff
@@ -771,22 +783,110 @@ StatusOr<RewriteObjectResponse> RestClient::RewriteObject(
...
+  // default (at least in this case), and that wastes bandwidth as the content
+  // length is known.
+  builder.AddHeader("Transfer-Encoding", {});
+
+  auto failure_predicate = [](rest::HttpStatusCode code) {
+    return (code != rest::HttpStatusCode::kResumeIncomplete &&
+            code >= rest::HttpStatusCode::kMinNotSuccess);
+  };
+
+  auto payload = request.payload();
```
Comment (coryan) [inline google/cloud/storage/internal/rest_client.cc:886]: I think this should be:

```cc
  return ParseFromRestResponse<QueryResumableUploadResponse>(
    storage_rest_client_->Put(std::move(builder).BuildRequest(), request.payload()),
    failure_predicate);
```

File: google/cloud/storage/client.cc [hunk 2]
```diff
@@ -96,6 +97,9 @@ std::shared_ptr<internal::RawClient> Client::CreateDefaultInternalClient(
       rest_opts.set<rest::EnableCurlSigpipeHandlerOption>(
           opts.get<EnableCurlSigpipeHandlerOption>());
     }
+    rest_opts.set<rest::IgnoredHttpErrorCodes>(
```
Comment (coryan) [inline google/cloud/storage/client.cc:100]: Does this means the status codes get ignored for all requests?  How does it work for other requests then?  Maybe add a comment?

**Review by coryan — CHANGES_REQUESTED** [review]
A small optimization for the uploads (I think).

File: google/cloud/storage/internal/rest_client.cc [hunk 3]
```diff
@@ -59,16 +59,26 @@ std::string UrlEscapeString(std::string const& value) {
   return std::string(handle.MakeEscapedString(value).get());
 }
 
+bool IsMinNotSuccess(rest::HttpStatusCode code) {
+  return code >= rest::kMinNotSuccess;
+}
```
Comment (devbww) [inline google/cloud/storage/internal/rest_client.cc:64]: Nit: It seems strange for something called "**Is**MinNotSuccess" to be implemented as "**>**= kMinNotSuccess".

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9274)* status: 0 of 6 files reviewed, 3 unresolved discussions (waiting on @coryan and @devbww)

___
*[`google/cloud/storage/client.cc` line 100 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9274#-N4V7jtRMzr2Lojnd5h--r1-100:-N4Y8Ifx3DLtACyc3g0V:bs3mhdg) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c9d389a2a3d993a77778ee98e0d4ea418e7b8be2/google/cloud/storage/client.cc#L100)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Does this means the status codes get ignored for all requests?  How does it work for other requests then?  Maybe add a comment?
</blockquote></details>

Comment added.
___
*[`google/cloud/storage/internal/rest_client.cc` line 64 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9274#-N4V7jtRMzr2Lojnd5h2-r1-64:-N4Y8JqGDR0D4EHv5Hz_:b903xua) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c9d389a2a3d993a77778ee98e0d4ea418e7b8be2/google/cloud/storage/internal/rest_client.cc#L64)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Nit: It seems strange for something called "**Is**MinNotSuccess" to be implemented as "**>**= kMinNotSuccess".
</blockquote></details>

I'm open to naming suggestions.
___
*[`google/cloud/storage/internal/rest_client.cc` line 886 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9274#-N4V7jtRMzr2Lojnd5h2-r1-886:-N4Y8MBV4uytpbGO41kJ:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c9d389a2a3d993a77778ee98e0d4ea418e7b8be2/google/cloud/storage/internal/rest_client.cc#L886)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think this should be:

```cc
  return ParseFromRestResponse<QueryResumableUploadResponse>(
    storage_rest_client_->Put(std::move(builder).BuildRequest(), request.payload()),
    failure_predicate);
```

</blockquote></details>

Done.

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 4 of 6 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9274)* status: all files reviewed, 2 unresolved discussions (waiting on @coryan, @devbww, and @scotthart)

___
*[`google/cloud/storage/internal/rest_client.cc` line 64 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9274#-N4V7jtRMzr2Lojnd5h2-r1-64:-N4YAWBJ-OaeVoeZtGER:b-4xmcob) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c9d389a2a3d993a77778ee98e0d4ea418e7b8be2/google/cloud/storage/internal/rest_client.cc#L64)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

I'm open to naming suggestions.
</blockquote></details>

Maybe `IsHttpFailure()`?
___
*[`google/cloud/storage/internal/rest_client.cc` line 886 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9274#-N4V7jtRMzr2Lojnd5h2-r1-886:-N4YAf_Q0AOgS8pV4DmS:bb5muef) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c9d389a2a3d993a77778ee98e0d4ea418e7b8be2/google/cloud/storage/internal/rest_client.cc#L886)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I think you do not want the extra braces around `request.payload()`, that makes an extra copy.  The `Put()` call consumes a `std::vector<absl::Span<char const>> const&`:

https://github.com/googleapis/google-cloud-cpp/blob/dab856b1517efc7a5759f78333c1470b2068e06f/google/cloud/internal/rest_client.h#L62-L64

the `.payload()` call returns the same type (modulo some type aliases):

https://github.com/googleapis/google-cloud-cpp/blob/dab856b1517efc7a5759f78333c1470b2068e06f/google/cloud/storage/internal/object_requests.h#L431

https://github.com/googleapis/google-cloud-cpp/blob/dab856b1517efc7a5759f78333c1470b2068e06f/google/cloud/storage/internal/const_buffer.h#L29-L33

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9274)* status: 5 of 6 files reviewed, 2 unresolved discussions (waiting on @coryan and @devbww)

___
*[`google/cloud/storage/internal/rest_client.cc` line 64 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9274#-N4V7jtRMzr2Lojnd5h2-r1-64:-N4YJEnn16utqePZOkeA:bskj59f) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c9d389a2a3d993a77778ee98e0d4ea418e7b8be2/google/cloud/storage/internal/rest_client.cc#L64)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Maybe `IsHttpFailure()`?
</blockquote></details>

Went with IsHttpError
___
*[`google/cloud/storage/internal/rest_client.cc` line 886 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9274#-N4V7jtRMzr2Lojnd5h2-r1-886:-N4YJIU61sPcCV1RPPBs:bv99wa2) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c9d389a2a3d993a77778ee98e0d4ea418e7b8be2/google/cloud/storage/internal/rest_client.cc#L886)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think you do not want the extra braces around `request.payload()`, that makes an extra copy.  The `Put()` call consumes a `std::vector<absl::Span<char const>> const&`:

https://github.com/googleapis/google-cloud-cpp/blob/dab856b1517efc7a5759f78333c1470b2068e06f/google/cloud/internal/rest_client.h#L62-L64

the `.payload()` call returns the same type (modulo some type aliases):

https://github.com/googleapis/google-cloud-cpp/blob/dab856b1517efc7a5759f78333c1470b2068e06f/google/cloud/storage/internal/object_requests.h#L431

https://github.com/googleapis/google-cloud-cpp/blob/dab856b1517efc7a5759f78333c1470b2068e06f/google/cloud/storage/internal/const_buffer.h#L29-L33


</blockquote></details>

Correct. I do not want to make an unnecessary copy. Thanks.

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9274#-:-N4YKbLQBv3SSjhuD5Xn:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @scotthart)

**Pull request merged**